osu.Server.BeatmapSubmission/BeatmapSubmissionController.cs

```diff
@@ -36,34 +37,47 @@ public BeatmapSubmissionController(IBeatmapStorage beatmapStorage, BeatmapPackag
+        /// <response code="422">The request was incorrectly formed, or could not be serviced due to violated invariants. Check returned error for details.</response>
         [HttpPut]
         [Route("beatmapsets")]
         [Consumes("application/json")]
         [Produces("application/json")]
         [ProducesResponseType(typeof(PutBeatmapSetResponse), 200)]
+        [ProducesResponseType(typeof(ErrorResponse), 422)]
         public async Task<IActionResult> PutBeatmapSetAsync([FromBody] PutBeatmapSetRequest request)
         {
+            uint userId = User.GetUserId();
+
             using var db = DatabaseAccess.GetConnection();
-            using var transaction = await db.BeginTransactionAsync();
 
-            // TODO: check silence state (https://github.com/ppy/osu-web/blob/master/app/Models/User.php#L1114-L1127)
-            // TODO: check restriction state (`SELECT user_warnings FROM phpbb_users WHERE user_id = $userId`)
-            // TODO: check difficulty limits (1 min, 128 max)
-            // TODO: check playcount (`("SELECT sum(playcount) FROM osu_user_month_playcount WHERE user_id = $userId") < 5`)
-            // TODO: clean up user's inactive maps
-            // TODO: check remaining map quota
+            ErrorResponse? userError = await checkUserAccountStanding(db, userId);
+            if (userError != null)
+                return userError.ToActionResult();
+
             // TODO: create forum thread for description editing purposes if set is new
 
-            uint userId = User.GetUserId();
+            if (await db.GetUserMonthlyPlaycountAsync(userId) < 5)
+                return new ErrorResponse("Thanks for your contribution, but please play the game first!").ToActionResult();
 
             uint? beatmapSetId = request.BeatmapSetID;
             uint[] existingBeatmaps = [];
 
+            using var transaction = await db.BeginTransactionAsync();
+
+            await db.PurgeInactiveBeatmapSetsForUserAsync(userId, transaction);
+            (uint totalSlots, uint remainingSlots) = await getUploadQuota(db, userId, transaction);
+
             if (beatmapSetId == null)
             {
                 if (request.BeatmapsToKeep.Length != 0)
-                    return UnprocessableEntity("Cannot specify beatmaps to keep when creating a new beatmap set.");
+                    return new ErrorResponse("Cannot specify beatmaps to keep when creating a new beatmap set.").ToActionResult();
+
+                if (remainingSlots <= 0)
+                {
+                    return new ErrorResponse($"You have exceeded your submission cap (you are currently allowed {totalSlots} total unranked maps). "
+                                             + $"Please finish the maps you currently have submitted, or wait until your submissions expire automatically to the graveyard "
+                                             + $"(about 4 weeks).").ToActionResult();
```

Comment:
```suggestion
                                             + $"(about 4 weeks since last update).").ToActionResult();
```

probably make this more concise. i guess eventually we could give a more definitive answer too.

Reply:
Self applying this to keep reviewing easily.